Class-based resource

keren-elbaz-m opened the pull request:
created generic base controller, service, and repository

keren-elbaz-m committed:
Class-based resource

created generic base controller, service, and repository

DanielBaronMoveo commented on src/modules/base/base.controller.ts:
```diff
@@ -0,0 +1,68 @@
+        } catch(e){
+            res.status(404).json({ message: 'Failed to retrieve item',e});
+        }
+    };
```
- Empty arrays are valid responses. This will incorrectly return 404 for empty collections. Should return 200 with empty array.
- In the catch block, we'd rather showcase a server error instead of not-found 404 one.
```
    getAll = async (req: Request, res: Response): Promise<void> => {
        try {
            const data = await this.service.getAll();
            res.json(data);
        } catch (e) {
            res.status(500).json({ message: 'Failed to retrieve items', error: e.message });
        }
    };

DanielBaronMoveo replied:
Same regarding the raw error object.

DanielBaronMoveo replied:
Taking the getAll as example, I'd mainly make an organize structure for my response and make it look very descriptive to handle with future usages: 

```
getAll = async (req: Request, res: Response): Promise<void> => {
        try {
            const data = await this.service.getAll();
            res.json({
                success: true,
                data,
                count: data.length
            });
        } catch (error) {
            console.error('Error in getAll:', error);
            res.status(500).json({
                success: false,
                message: 'Failed to retrieve items',
                error: error.message
            });
        }
    };

DanielBaronMoveo commented on src/modules/base/base.controller.ts:
```diff
@@ -0,0 +1,68 @@
+        } catch(e){
+            res.status(500).json({ message: 'Failed to create item', e });
+        }
+    };
```
- Wrong error message (should be "Failed to retrieve item", not "create")
- Exposing raw error object e

```
getById = async (req: Request, res: Response): Promise<void> => {
        try {
            const data = await this.service.getById(req.params.id);
            if (!data) {
                res.status(404).send('Not found');
                return;
            }
            res.json(data);
        } catch (e) {
            res.status(500).json({ message: 'Failed to retrieve item', error: e.message });
        }
    };

DanielBaronMoveo commented on src/modules/base/base.controller.ts:
```diff
@@ -0,0 +1,68 @@
+        } catch (e) {
+            res.status(500).json({ message: 'Failed to create item', e });
+        }
+    };
```
- Exposing a raw error object instead of it's message.

```
 create = async (req: Request, res: Response): Promise<void> => {
        try {
            const created = await this.service.create(req.body as T);
            res.status(201).json(created);
        } catch (e) {
            res.status(500).json({ message: 'Failed to create item', error: e.message });
        }
    };

DanielBaronMoveo commented on src/modules/base/base.service.ts:
```diff
@@ -0,0 +1,27 @@
+    constructor(private repo: BaseRepository<T>) {
+    }
+
+    async getAll(): Promise<T[]> {
+        return this.repo.findAll();
+    }
+
+    async getById(id: string): Promise<T | null> {
+        return this.repo.findById(id);
+    }
```
Consistent method naming - changed repository calls from findAll()/findById() to getAll()/getById() for consistency

DanielBaronMoveo commented on src/modules/base/base.repository.ts:
```diff
@@ -0,0 +1,26 @@
+import { Model, Document } from "mongoose";
+
+export class BaseRepository <T extends Document>{
+    constructor(private model:Model<T>){}
```
Add readonly modifier - prevents accidental reassignment and maintains consistency

DanielBaronMoveo commented on src/modules/base/base.service.ts:
```diff
@@ -0,0 +1,27 @@
+import { BaseRepository } from './base.repository';
+
+export class BaseService<T extends Document> {
+    constructor(private repo: BaseRepository<T>) {
```
Add readonly modifier - prevents accidental reassignment and maintains consistency

DanielBaronMoveo commented on src/modules/base/base.service.ts:
```diff
@@ -0,0 +1,27 @@
+        return this.repo.create(data);
+    }
+
+    async update(id: string, data: UpdateQuery<T>): Promise<T | null> {
+        return this.repo.update(id, data);
+    }
```
Using Partial<T> instead of UpdateQuery<T> which is what the service and controller expect. This creates type inconsistency.

Be consistent with the one you're using. Either Partial or UpdateQuery

DanielBaronMoveo reviewed (changes requested):
Good job, let's fix the required comments. After you're done with each comment please comment with a sign / "done" so I'll know you did care of it! 🥇

keren-elbaz-m committed:
Fixed PR comments

Unified error responses: return error.message instead of raw error object and improved controller responses.

Replaced Partial<T> with UpdateQuery<T> in service and repository for type consistency
